### PR TITLE
Add validator in rich text components for mention configuration and batch updates

### PR DESCRIPTION
**Pull request by vivinkrishna-ni**
# Pull Request

## 🤨 Rationale

1. Add a validator for rich text components to handle the duplicate configuration elements in their child and expose the validation APIs `checkValidity()` and `validity`. Related conversation: https://github.com/ni/nimble/pull/1664#discussion_r1401133453
2. As the `handleChange` for any changes in the configuration or mapping elements is called multiple times which eventually calls the `updateView()` to get and set the markdown in both editor and viewer. Since the parser and serializer are expensive operations, we should reduce the number of calls to update the view.

## 👩‍💻 Implementation

- Add a helper class `RichTextValidator` in the rich text base model folder to validate the inputs given by the user to configure the mentions. The validation includes:
    - `invalidMentionConfiguration`: which reflects the validation state of the configuration element
    - `duplicateMentionConfiguration`: validates if there is more than one identical configuration element within the children of the rich text editor/viewer. As the editor/viewer should not have multiple data for one mention type, we restrict through a validation.
- Expose `checkValidity()` and `validity` read-only APIs for the user to know what aspect of the mention configuration fails for the editor/viewer. Include the same in Angular directives too.
- Added `queueUpdate()` to update the view only when necessary in both editor and viewer components by utilizing the `DOM.queueUpdate` from FAST.

## 🧪 Testing

- Relied on the existing test cases for batch updates
- Updated a few test cases in the editor and viewer related to new validation additon
- Added test cases to editor and viewer spec files for rich text validator

## ✅ Checklist

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

### CODE REVIEW

**Review by m-akinc — APPROVED** [review]

File: angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts
Comment (vivinkrishna-ni) [inline angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts:1]: Reviewer's info: Having a `RichText` base class just like the nimble-components will be taken in a follow-up PR.

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 1]
```diff
@@ -36,22 +66,37 @@ export abstract class RichText extends FoundationElement {
...
+    protected mentionElementsChanged(): void {
+        if (!this.$fastController.isConnected) {
+            return;
+        }
+
+        this.richTextValidator.validateDuplicateMentionConfigurations(
+            this.mentionElements
+        );
+        this.observeMentions();
+        if (this.mentionElements.length) {
```
[the remark this answers is not in the document]
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/index.ts]: Updated `#2` and `#3` as suggested.

Regarding `#1`, I agree that if the `mentionElements` is emptied, we should update the view. However, when the element is dynamically added/removed by the client, the `handleChange()` in the same file will be called with the change args as `mentionConfigs` which calls the `queueUpdate` to update the view.

So, I would like to understand whether it is necessary to update the view during the component initialization, even if the `mentionElements` are not available. This consideration arises due to the involvement of the `updateView()` method, which includes getMarkdown + setMarkdown. Due to this, I see the following drawbacks:
1. There are some [existing async test failures](https://github.com/ni/nimble/actions/runs/7018282186/job/19093340940) in the editor when no child items (mentionElements) are present via the template. I tried adding a `isConnected` condition to `childItemsChanged` but that doesn't call the update method as the editor is not yet connected. Some of the test cases are resolved once after adding `waitForUpdatesAsync` next to `await connect()` as mentioned in [the thread](https://github.com/ni/nimble/pull/1099/files#r1130001024) which I referred the patter from [table batch updates PR](https://github.com/ni/nimble/pull/1099/files).
2. As mentioned, `updateView()` involves getMarkdown + setMarkdown operations each time it is called, making it quite an expensive operation involving both the parser and serializer. So the clients who don't need the mention configuration (only basic formatting options) also will go through this operation for the first time during initialization which seems unnecessary.

Due to the above reasons, we thought of having a condition to update only if the allowed configuration elements are added as children. In the case of adding/removing elements, the `handleChange` will be responsible for updating the view. @m-akinc Please let me know your thoughts on whether we could use this condition here to resolve the above issues or let me know if I missed some other case upon adding this condition.

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 2]
```diff
@@ -36,22 +66,37 @@ export abstract class RichText extends FoundationElement {
 
     protected abstract updateView(): void;
 
+    protected mentionElementsChanged(): void {
+        if (!this.$fastController.isConnected) {
```
[the remark this answers is not in the document]
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/index.ts]: It makes sense to not check the `isConnected` here and I ensured that the validator is initialized before this connection so without this it worked as expected.

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 3]
```diff
@@ -16,16 +23,39 @@ export abstract class RichText extends FoundationElement {
...
+        return this.richTextValidator.isValid();
+    }
 
     /**
      * @internal
      */
     public handleChange(source: unknown, args: unknown): void {
         if (source instanceof MentionInternals && typeof args === 'string') {
-            this.updateMentionConfig();
+            if (args === 'validConfiguration') {
```
[the remark this answers is not in the document]
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/index.ts:51]: I agree. Updated as suggested.

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 4]
```diff
@@ -89,4 +132,18 @@ export abstract class RichText extends FoundationElement {
         });
         this.mentionInternalsNotifiers = [];
     }
+
+    private queueUpdate(): void {
+        if (!this.$fastController.isConnected) {
```
[the remark this answers is not in the document]
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/index.ts]: Removed this check

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 5]
```diff
@@ -89,4 +132,18 @@ export abstract class RichText extends FoundationElement {
...
+            return;
+        }
+
+        if (!this.updateQueued) {
+            this.updateQueued = true;
+            DOM.queueUpdate(() => {
+                this.updateMentionConfig();
+                this.updateQueued = false;
```
[the remark this answers is not in the document]
Comment (atmgrifter00) [reply packages/nimble-components/src/rich-text/base/index.ts:123]: I mean, this is still a single-threaded process running correct? I don't think it's possible for a separate thread to run the code in between the two lines that are contained within the provided `queueUpdate` implementation.

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts [hunk 6]
```diff
@@ -2112,20 +2112,13 @@ describe('RichTextEditor', () => {
...
 
-            expect(pageObject.getMarkdownRenderedTagNames()).toEqual([
-                'P',
-                `${richTextMentionUsersViewTag}`.toUpperCase(),
-                'BR'
-            ]);
-            expect(
-                pageObject.getEditorMentionViewAttributeValues('mention-label')
-            ).toEqual(['1']);
+            expect(pageObject.getEditorTagNames()).toEqual(['P', 'A']);
```
Comment (atmgrifter00) [inline packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:2120]: Why is this test not using the `anchorTag` approach as done in the rich-text-editor tests? Are we using different elements between the two?

**Review by atmgrifter00 — APPROVED** [review]

File: packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts (hunk 6 above)
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:2120]: Yes, we are using `anchorTag` in the viewer but not in the editor due to the issues listed in https://github.com/ni/nimble/issues/1516. These issues are majorly related to rendering `nimble-anchor` in `contenteditable` div like the rich text editor. Whereas in the viewer, it does not have any issues in rendering so used the same. The goal is to use `nimble-anchor` in both components and in the editor, we will replace `<a>` with `nimble-anchor` once the linked issue is resolved.

File: angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts [hunk 7]
```diff
@@ -60,6 +62,10 @@ export class NimbleRichTextEditorDirective {
         this.renderer.setProperty(this.elementRef.nativeElement, 'placeholder', value);
     }
 
+    public get validity(): RichTextValidity {
```
Comment (jattasNI) [inline angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts:65]: Updates to directives should have corresponding updates to their spec files

File: packages/nimble-components/src/rich-text/base/models/rich-text-validator.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import type { RichTextMention } from '../../../rich-text-mention/base';
+import type { RichTextValidity } from '../types';
+
+/**
+ * Helper class for the nimble rich text components to validate that the configuration
+ * is valid and report which aspects of the configuration are valid or invalid.
+ */
+export class RichTextValidator {
```
Comment (jattasNI) [inline packages/nimble-components/src/rich-text/base/models/rich-text-validator.ts:8]: For each of the invalid cases we should have a test coverage that ensure `isValid()` is `false` and `getValidity()` returns an object with the right field set. You can look at the table validator spec as an example.

File: packages/nimble-components/src/rich-text/base/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-akinc) [reply packages/nimble-components/src/rich-text/base/index.ts]: As was discussed in a meeting, the plan is to try removing the condition from `updateMentionConfig` and adding one to `mentionElementsChanged` that only skips queuing the update when the value goes from `undefined` to `[]`.

**Review by jattasNI — COMMENTED** [review]
Once my feedback is addressed and other reviewers have approved, no need to wait for me to submit since I'm OOO this week.

File: packages/nimble-components/src/rich-text/base/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (m-akinc) [reply packages/nimble-components/src/rich-text/base/index.ts:126]: > this is still a single-threaded process running correct?

I didn't know. But looking into it, that sounds like it's essentially true. JavaScript does support worker threads, but it sounds like that's not common (we're certainly not using them). So never mind my concern!

File: packages/nimble-components/src/rich-text/base/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/index.ts]: @m-akinc Sorry if I didn't understand it properly during the call. I think it is better to do the check in the `mentionElementsChanged` to skip the queuing when it goes from undefined to `[]`. However, `#2` in the first comment of this thread, you mentioned that `mentionElements` need not be observable as it is assigned a value only in the `childItemsChanged` method. I hope earlier we don't have a proper reason to have it but now we have a purpose of checking the old and new state so we can use an observable for `mentionElements`. Please correct me if my understanding is wrong.

> the plan is to try removing the condition from `updateMentionConfig`

However, we discussed removing the check from `updateMentionsFromChildItems` and not from `updateMentionConfig` which is responsible for updating the parser config and updating the view by checking whether the given configuration is valid or not.

File: angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts (hunk 7 above)
Comment (vivinkrishna-ni) [reply angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts:65]: I have added `validity` and `checkValidity` default value test cases to the spec files.

File: packages/nimble-components/src/rich-text/base/models/rich-text-validator.ts (hunk 8 above)
Comment (vivinkrishna-ni) [reply packages/nimble-components/src/rich-text/base/models/rich-text-validator.ts:8]: Covered the test cases in editor and viewer spec files

File: packages/nimble-components/src/rich-text/base/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-akinc) [reply packages/nimble-components/src/rich-text/base/index.ts]: Sorry, I misspoke in my comment when I said we would be adding the check to `mentionElementsChanged`. I meant `childItemsChanged`. So this is what I had expected:
```TS
private childItemsChanged(prev: Element[] | undefined, next: Element[]): void {
    if (prev?.length || next.length) {
        void this.updateMentionsFromChildItems();
    }
}

private async updateMentionsFromChildItems(): Promise<void> {
    const definedElements = this.childItems.map(async item => (item.matches(':not(:defined)')
        ? customElements.whenDefined(item.localName)	
         : Promise.resolve()));
    await Promise.all(definedElements);
    this.mentionElements = this.childItems.filter(
        (x): x is RichTextMention => x instanceof RichTextMention
    );
    this.observeMentions();
    this.queueUpdate();
}
```
Then we still wouldn't need to define `mentionElementsChanged` and `mentionElements` wouldn't need to be `@observable`.

File: packages/nimble-components/src/rich-text/base/index.ts [hunk 9]
```diff
@@ -41,24 +65,35 @@ export abstract class RichText extends FoundationElement {
...
         this.updateView();
     }
 
     private childItemsChanged(): void {
         void this.updateMentionsFromChildItems();
     }
 
+    private mentionElementsChanged(
+        prev: string[] | undefined,
+        next: string[]
```
Comment (m-akinc) [inline packages/nimble-components/src/rich-text/base/index.ts:87]: As I responded in the other thread, I don't think this function is actually needed, but if for whatever reason it is, the types of `prev` and `next` should be `RichTextMention[]`, not `string[]`.

File: angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts [hunk 10]
```diff
@@ -1,8 +1,10 @@
 import { Directive, ElementRef, EventEmitter, HostListener, Input, Output, Renderer2 } from '@angular/core';
```
Comment (rajsite) [inline angular-workspace/projects/ni/nimble-angular/rich-text/editor/nimble-rich-text-editor.directive.ts:1]: There are a large number of changes in #1683 that have not been merged here yet. I think I need to wait for that merge for review.

**Comment by vivinkrishna-ni** [comment]
We have created a new PR with the same changes and addressed the open comments there: https://github.com/ni/nimble/pull/1744. So I am closing this PR.

**Pull request closed without merging**